File: plugins/harness/skills/harness-install/assets/maven/harness-maven-plugin/src/main/java/com/ririnto/sinon/harness/rules/text/HookCommandRule.java
package com.ririnto.sinon.harness.rules.text;

import com.ririnto.sinon.harness.rules.HarnessCheckHelper;
import com.ririnto.sinon.harness.rules.HarnessCheckRule;
import com.ririnto.sinon.harness.core.RuleContext;
import com.ririnto.sinon.harness.Finding;

import tools.jackson.databind.JsonNode;
import org.apache.maven.plugin.MojoExecutionException;
import java.nio.file.Path;
import java.util.Collection;
import java.util.List;
import java.util.Objects;
import java.util.regex.Pattern;
import java.util.stream.Stream;

/**
 * Rule that requires hook files to declare and run harness validation commands.
 */
public enum HookCommandRule implements HarnessCheckRule {
    INSTANCE;

    @Override
    public String category() {
        return "hookCommand";
    }
    private static final String CATEGORY = "hookCommand";

    @Override
    public boolean applies(RuleContext ctx) {
        return ctx.manifest().isEnabled(CATEGORY);
    }

    @Override
    public Collection<Finding> validate(RuleContext ctx) throws MojoExecutionException {
        final Path root = ctx.root();
        final JsonNode manifest = ctx.manifest().raw();
        final JsonNode catNode = manifest.get(CATEGORY);
        final JsonNode allowedCmds = catNode.get("parameters").get("allowedCommands").get("maven");
        final String expectedCmd = allowedCmds == null || allowedCmds.size() == 0 ? "" : allowedCmds.get(0).asText();
        final String severity = HarnessCheckHelper.getSeverity(manifest, CATEGORY);
        final String prePushPath = catNode.get("parameters").get("prePushHook").asText();
        final Path prePush = root.resolve(prePushPath);
        final List<Finding> prePushIssues = Stream.of(prePush)
                .filter(p -> HarnessCheckHelper.isSafeRegularFile(root, p))
                .flatMap(p -> {
                    try {
                        final String prePushText = HarnessCheckHelper.readFile(root, p);
                        final String declaredCmd = extractHookCommand(prePushText);
                        return Stream.of(declaredCmd.isEmpty()
                                ? Finding.of(severity, CATEGORY, "pre-push hook must declare Harness validation command")
                                : !expectedCmd.equals(declaredCmd)
                                ? Finding.of(severity, CATEGORY, "pre-push hook declares unsupported validation command: " + declaredCmd)
                                : !prePushText.contains(declaredCmd)
                                ? Finding.of(severity, CATEGORY, "pre-push hook must run the declared validation command")
                                : null);
                    } catch (MojoExecutionException e) {
                        return Stream.empty();
                    }
                })
                .filter(Objects::nonNull)
                .toList();
        final String preCommitPath = catNode.get("parameters").get("preCommitHook").asText();
        final Path preCommit = root.resolve(preCommitPath);
        final List<Finding> preCommitIssues = validatePreCommitHook(root, severity, preCommit);
        return Stream.concat(prePushIssues.stream(), preCommitIssues.stream()).toList();
    }

    private String extractHookCommand(String text) {
        return Stream.of(text.split("\\R"))
                .filter(line -> line.startsWith("# Harness validation command: "))
                .map(line -> line.substring("# Harness validation command: ".length()).trim())
                .findFirst()
                .orElse("");
    }

    private List<Finding> validatePreCommitHook(Path root, String severity, Path preCommit) throws MojoExecutionException {
        return Stream.of(preCommit)
                .filter(p -> HarnessCheckHelper.isSafeRegularFile(root, p))
                .flatMap(p -> {
                    try {
                        final String preCommitText = HarnessCheckHelper.readFile(root, p);
                        final Pattern fullStackPattern = Pattern.compile("(^|\\s)(uv|bun|gradle|mvn)(\\s|$)|\\.\\./gradlew|harnessValidate|harness_validate\\.py|harness-validate\\.ts");
                        return fullStackPattern.matcher(preCommitText).find()
                                ? Stream.of(Finding.of(severity, CATEGORY, "pre-commit hook must not run full stack validation commands"))
                                : Stream.empty();
                    } catch (MojoExecutionException e) {
                        return Stream.empty();
                    }
                })
                .toList();
    }
}

File: plugins/harness/skills/harness-install/assets/gradle/buildSrc/src/main/kotlin/ai/harness/gradle/rules/HookCommandRule.kt
package ai.harness.gradle.rules

import ai.harness.gradle.HarnessPsiResults.Finding
import ai.harness.gradle.HarnessCheck
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import java.nio.file.Path
import kotlin.io.path.div
import kotlin.io.path.isRegularFile
import kotlin.io.path.readText

/**
 * Rule that requires hooks to declare and run validation commands.
 */
object HookCommandRule : HarnessCheckRule() {
    override val category: String = "hookCommand"
    override fun applies(manifest: JsonObject): Boolean =
        manifest[category]?.jsonObject?.get("enabled")?.jsonPrimitive?.contentOrNull?.toBoolean() ?: true

    override fun validate(
        manifest: JsonObject,
        root: Path,
    ): Collection<Finding> {
        val severity = HarnessCheck.severityOf(manifest, category)
        val catObj = manifest[category]?.jsonObject
        val parametersObj = catObj?.get("parameters")?.jsonObject
        val messagesObj = catObj?.get("messages")?.jsonObject
        return when {
            catObj == null || parametersObj == null || messagesObj == null -> {
                emptyList()
            }

            else -> {
                val allowedCmdObj = parametersObj["allowedCommands"]?.jsonObject
                val allowedCmds = HarnessCheck.stringArrayFrom(allowedCmdObj, "gradle")
                val allowedPreCommitCmdObj = parametersObj["allowedPreCommitCommands"]?.jsonObject
                val allowedPreCommitCmds = HarnessCheck.stringArrayFrom(allowedPreCommitCmdObj, "gradle")
                val prePushPath = HarnessCheck.stringFrom(parametersObj, "prePushHook")
                val preCommitPath = HarnessCheck.stringFrom(parametersObj, "preCommitHook")
                val prePushHook = root / prePushPath
                val preCommitHook = root / preCommitPath
                val prePushFindings =
                    if (prePushHook.isRegularFile()) {
                        val text = prePushHook.readText()
                        val command =
                            text
                                .lineSequence()
                                .firstOrNull { line ->
                                    line.startsWith("# Harness validation command: ")
                                }?.removePrefix("# Harness validation command: ")
                                ?.trim()
                                ?: ""
                        buildList {
                            if (command.isEmpty()) {
                                add(
                                    Finding(
                                        severity,
                                        category,
                                        HarnessCheck.stringFrom(messagesObj, "missingDeclaration").takeIf { message ->
                                            message.isNotEmpty()
                                        } ?: "pre-push hook must declare Harness validation command",
                                    ),
                                )
                            }
                            if (command.isNotEmpty() && command !in allowedCmds) {
                                add(
                                    Finding(
                                        severity,
                                        category,
                                        HarnessCheck.stringFrom(messagesObj, "unsupportedCommand").takeIf { message ->
                                            message.isNotEmpty()
                                        } ?: "pre-push hook declares unsupported validation command: $command",
                                    ),
                                )
                            }
                            if (command.isNotEmpty() && !text.contains(command)) {
                                add(
                                    Finding(
                                        severity,
                                        category,
                                        HarnessCheck.stringFrom(messagesObj, "commandNotRun").takeIf { message ->
                                            message.isNotEmpty()
                                        } ?: "pre-push hook must run the declared validation command",
                                    ),
                                )
                            }
                        }
                    } else {
                        emptyList()
                    }
                val preCommitFindings =
                    if (preCommitHook.isRegularFile()) {
                        val text = preCommitHook.readText()
                        val command =
                            text
                                .lineSequence()
                                .firstOrNull { line ->
                                    line.startsWith("# Harness validation command: ")
                                }?.removePrefix("# Harness validation command: ")
                                ?.trim()
                                ?: ""
                        buildList {
                            if (command.isEmpty() && allowedPreCommitCmds.isNotEmpty()) {
                                add(
                                    Finding(
                                        severity,
                                        category,
                                        HarnessCheck.stringFrom(messagesObj, "missingDeclaration").takeIf { message ->
                                            message.isNotEmpty()
                                        } ?: "pre-commit hook must declare validation command",
                                    ),
                                )
                            }
                            if (command.isNotEmpty() && command !in allowedPreCommitCmds) {
                                add(
                                    Finding(
                                        severity,
                                        category,
                                        HarnessCheck.stringFrom(messagesObj, "unsupportedCommand").takeIf { message ->
                                            message.isNotEmpty()
                                        } ?: "pre-commit hook declares unsupported validation command: $command",
                                    ),
                                )
                            }
                            if (command.isNotEmpty() && !text.contains(command)) {
                                add(
                                    Finding(
                                        severity,
                                        category,
                                        HarnessCheck.stringFrom(messagesObj, "commandNotRun").takeIf { message ->
                                            message.isNotEmpty()
                                        } ?: "pre-commit hook must run the declared validation command",
                                    ),
                                )
                            }
                        }
                    } else {
                        emptyList()
                    }
                prePushFindings + preCommitFindings
            }
        }
    }

    override fun renderPsiFindings(
        findings: List<ai.harness.gradle.PsiFinding>,
        manifest: JsonObject,
    ): Collection<Finding> = emptyList()

    override fun findPsiFindings(
        file: Path,
        root: Path,
        psiFactory: org.jetbrains.kotlin.psi.KtPsiFactory?,
    ): List<ai.harness.gradle.PsiFinding> = emptyList()
}
